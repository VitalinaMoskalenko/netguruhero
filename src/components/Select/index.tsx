import React, { useState } from "react";
import styled from "styled-components";
import { Body, SmallBody } from "../Paragraphs";

type PropsType = {
  emptyText: string;
  value?: string;
  label: string;
  data: string[];
  onValueSelect: (value: string) => void;
  errorMessage?: string;
};

const Container = styled.div`
  position: relative;
`;

type ValueContainerType = {
  isOpen: boolean;
  isError: boolean;
};

const ValueContainer = styled.div<ValueContainerType>`
  cursor: pointer;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: ${({ theme }) => theme.colors.white};
  border-radius: 8px;
  height: 40px;
  border: ${({ isError, theme }) =>
    isError ? `1px solid ${theme.colors.red}` : "0px"};
  padding: 0px 16px;
  border-bottom-left-radius: ${({ isOpen }) => (isOpen ? 0 : 8)}px;
  border-bottom-right-radius: ${({ isOpen }) => (isOpen ? 0 : 8)}px;
`;

const ValueListContainer = styled.div`
  position: absolute;
  top: 65px;
  bottom: 0;
  left: 0;
  right: 0;
  border-bottom-left-radius: 8px;
  border-bottom-right-radius: 8px;
  overflow: auto;
  height: 200px;
  background-color: ${({ theme }) => theme.colors.white};
  z-index: 100;
`;

const DropDownIcon = styled(Body)`
  color: ${({ theme }) => theme.colors.black05};
  transform: rotate(270deg);
`;

const Label = styled(SmallBody)`
  font-weight: bold;
  margin-bottom: 6px;
  opacity: 0.5;
`;

const ErrorMessage = styled(SmallBody)`
  margin-top: 4px;
  color: ${({ theme }) => theme.colors.red};
`;

const ValueButtonContainer = styled.button`
  cursor: pointer;
  text-align: left;
  width: 100%;
  height: 40px;
  border: 0px;
  padding: 0px 16px;
  border-top: 1px solid ${({ theme }) => theme.colors.gray};
  background-color: ${({ theme }) => theme.colors.white};
  font-family: "Courier New", Courier, monospace;

  :hover {
    background-color: ${({ theme }) => theme.colors.gray};
  }
`;

const Select = ({
  label,
  data,
  value,
  emptyText,
  onValueSelect,
  errorMessage,
}: PropsType) => {
  const [isOpen, setIsOpen] = useState(false);
  const [selectedValue, setSelectedValue] = useState(value);

  const toggleIsOpen = () => {
    setIsOpen(!isOpen);
  };

  const selectValue = (item: string) => {
    setSelectedValue(item);
    onValueSelect(item);
    toggleIsOpen();
  };

  return (
    <Container>
      <Label>{label}</Label>
      <ValueContainer
        onClick={toggleIsOpen}
        isOpen={isOpen}
        isError={!!errorMessage}
      >
        <Body>{selectedValue || emptyText}</Body>
        <DropDownIcon>&#10094;</DropDownIcon>
      </ValueContainer>
      {errorMessage && <ErrorMessage>{errorMessage}</ErrorMessage>}
      {isOpen && (
        <ValueListContainer>
          {data.map((item, index) => {
            return (
              <ValueButtonContainer
                onClick={() => selectValue(item)}
                key={item}
              >
                <Body>{item}</Body>
              </ValueButtonContainer>
            );
          })}
        </ValueListContainer>
      )}
    </Container>
  );
};

export default Select;
