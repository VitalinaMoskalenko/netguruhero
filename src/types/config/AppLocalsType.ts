export enum AppLocalsType {
  en = "en",
}
