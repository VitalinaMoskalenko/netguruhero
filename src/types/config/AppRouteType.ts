export enum AppRouteType {
  home = "/",
  addHeroForm = "/addHeroForm",
  heroDetails = "/heroDetails",
  heroDetailsWithId = "/heroDetails/:id",
}
