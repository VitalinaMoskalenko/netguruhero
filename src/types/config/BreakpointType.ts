export enum BreakpointType {
  xxs = 375,
  xs = 768,
  xxl = 1440,
}
