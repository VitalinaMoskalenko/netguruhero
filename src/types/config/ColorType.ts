export enum ColorType {
  white = "white",
  black = "black",
  green = "#63cd93",
  blue = "#6199f3",
  gray = "#f5f6fb",
  black05 = "rgba(0,0,0, 0.5)",
}
