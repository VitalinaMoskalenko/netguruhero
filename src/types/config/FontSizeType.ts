export enum FontSizeType {
  smallBody = 14,
  body = 18,
  h5 = 16,
  h4 = 18,
  h3 = 24,
  h2 = 32,
  h1 = 48,
}
