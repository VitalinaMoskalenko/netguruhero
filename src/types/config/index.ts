export * from "./AppRouteType";
export * from "./FontSizeType";
export * from "./ColorType";
export * from "./BreakpointType";
export * from "./AppLocalsType";
