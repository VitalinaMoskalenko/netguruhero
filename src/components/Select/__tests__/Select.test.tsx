import React from "react";
import renderer from "react-test-renderer";
import { ThemeProvider } from "styled-components";
import Select from "../";
import { theme } from "../../../config";

it("renders select correctly", () => {
  const selectComponent = renderer
    .create(
      <ThemeProvider theme={theme}>
        <Select
          label=""
          data={[]}
          value=""
          emptyText=""
          onValueSelect={() => {}}
        />
      </ThemeProvider>
    )
    .toJSON();

  expect(selectComponent).toMatchSnapshot();
});
