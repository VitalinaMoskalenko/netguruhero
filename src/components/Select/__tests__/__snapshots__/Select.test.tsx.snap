// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders select correctly 1`] = `
<div
  className="sc-dlnjwi fHmCXS"
>
  <p
    className="sc-gtsrHT sc-gKAaRy ckjQab gALeJx"
  >
    
  </p>
  <div
    className="sc-hKFxyN yeSJZ"
    onClick={[Function]}
  >
    <p
      className="sc-bdnxRM jupkGm"
    >
      
    </p>
    <p
      className="sc-bdnxRM sc-jSFjdj jupkGm gdsnfc"
    >
      ‚ùÆ
    </p>
  </div>
</div>
`;
